File: samples/cext-trey-research/src/functions/me.ts
/* This code sample provides a starter kit to implement server side logic for your Teams App in TypeScript,
 * refer to https://docs.microsoft.com/en-us/azure/azure-functions/functions-reference for complete Azure Functions
 * developer guide.
 */

import { app, HttpRequest, HttpResponseInit, InvocationContext } from "@azure/functions";
import ConsultantApiService from "../services/ConsultantApiService";
import { ApiConsultant, ApiChargeTimeResponse, ErrorResult } from "../model/apiModel";
import { HttpError, cleanUpParameter } from "../services/Utilities";
import IdentityService from "../services/IdentityService";
/**
 * This function handles the HTTP request and returns my information.
 *
 * @param {HttpRequest} req - The HTTP request.
 * @param {InvocationContext} context - The Azure Functions context object.
 * @returns {Promise<HttpResponse>} - A promise that resolves with the HTTP response containing my information.
 */

// Define a Response interface.
interface Response extends HttpResponseInit {
  status: number;
  jsonBody: {
    results: ApiConsultant[] | ApiChargeTimeResponse | ErrorResult;
  };
}
export async function me(
  req: HttpRequest,
  context: InvocationContext
): Promise<Response> {
  context.log("HTTP trigger function me processed a request.");

  // Initialize response.
  const res: Response = {
    status: 200,
    jsonBody: {
      results: [],
    },
  };

  try {
    const me = await IdentityService.validateRequest(req);
    const command = req.params.command?.toLowerCase();   
    let body=null; 
    switch (req.method) {
      case "GET": {

        if (command) {
          throw new HttpError(400, `Invalid command: ${command}`);
        }

        console.log(`➡️ GET /api/me request`);

        const result = [ me ];
        res.jsonBody.results = result;
        console.log(`   ✅ GET /me response status ${res.status}; ${result.length} consultants returned`);
        return res;
      }
      case "POST": {
        const me = await IdentityService.validateRequest(req);
        switch (command) {
          case "chargetime": {  
            try {
              const bd = await req.text();
              body = JSON.parse(bd);
            } catch (error) {
              throw new HttpError(400, `No body to process this request.`);
            }
            if (body) {              
            
            const projectName = cleanUpParameter("projectName", body["projectName"]);
            if (!projectName) {
              throw new HttpError(400, `Missing project name`);
            }
            const hours = body["hours"];
            if (!hours) {
              throw new HttpError(400, `Missing hours`);
            }
            if (typeof hours !== 'number' || hours < 0 || hours > 24) {
              throw new HttpError(400, `Invalid hours: ${hours}`);
            }

            console.log(`➡️ POST /api/me/chargetime request for project ${projectName}, hours ${hours}`);
            const result = await ConsultantApiService.chargeTimeToProject (projectName, me.id, hours);

            res.jsonBody.results = {
              status: 200,
              clientName: result.clientName,
              projectName: result.projectName,
              remainingForecast: result.remainingForecast,
              message: result.message
            };
            console.log(`   ✅ POST /api/me/chargetime response status ${res.status}; ${result.message}`);
            } else {
              throw new HttpError(400, `Missing request body`);
            }
            return res;
          }
          default: {
            throw new HttpError(400, `Invalid command: ${command}`);
          }
        }
      }
      default:
        throw new HttpError(405, `Method not allowed: ${req.method}`);
    }

  } catch (error) {

    const status = <number>error.status || <number>error.response?.status || 500;
    console.log(`   ⛔ Returning error status code ${status}: ${error.message}`);

    res.status = status;
    res.jsonBody.results = {
      status: status,
      message: error.message
    };
    return res;
  }

}

app.http("me", {
  methods: ["GET","POST"],
  authLevel: "anonymous",
  route: "me/{*command}",
  handler: me,
});

File: samples/cext-trey-research-python/functions/me/me.py
import azure.functions as func
import logging
import json
import re
from services.identity_service import identity_service
from services.consultant_api_service import consultant_api_service
from services.utilities import HttpError, clean_up_parameter
import dataclasses

async def me(req: func.HttpRequest) -> func.HttpResponse:
    logging.info("HTTP trigger function 'me' processed a request.")

    res = {
        "status": 200,
        "results": []
    }

    try:
        me = await identity_service.validate_request(req)
        command = req.route_params.get("command", "").lower()
        body = None

        if req.method == "GET":
            if command:
                raise HttpError(400, f"Invalid command: {command}")

            logging.info("➡️ GET /api/me request")
            res["results"] = [dataclasses.asdict(me)]
            logging.info(f"✅ GET /me response status {res['status']}; {len(res['results'])} consultants returned")
            return func.HttpResponse(
                json.dumps(res),
                status_code=res["status"],
                mimetype="application/json"
            )

        elif req.method == "POST":
            if command == "chargetime":
                try:
                    raw_body = req.get_body().decode('utf-8')
                    formatted_body = re.sub(r'([a-zA-Z0-9_]+):', r'"\1":', raw_body)
                    body = json.loads(formatted_body)
                except ValueError as e:
                    logging.error(f"Failed to parse JSON body: {e}")
                    raise HttpError(400, "No body to process this request.")

                if body:
                    project_name = clean_up_parameter("projectName", body.get("projectName"))
                    if not project_name:
                        raise HttpError(400, "Missing project name")

                    hours = body.get("hours")
                    if not hours:
                        raise HttpError(400, "Missing hours")
                    if not isinstance(hours, (int, float)) or hours < 0 or hours > 24:
                        raise HttpError(400, f"Invalid hours: {hours}")

                    logging.info(f"➡️ POST /api/me/chargetime request for project {project_name}, hours {hours}")
                    result = await consultant_api_service.charge_time_to_project(project_name, me.id, hours)

                    res["results"] = {
                        "status": 200,
                        "clientName": result.clientName,
                        "projectName": result.projectName,
                        "remainingForecast": result.remainingForecast,
                        "message": result.message
                    }
                    logging.info(f"✅ POST /api/me/chargetime response status {res['status']}; {result.message}")

                else:
                    raise HttpError(400, "Missing request body")
                return func.HttpResponse(
                    json.dumps(res),
                    status_code=res["status"],
                    mimetype="application/json"
                )
            else:
                raise HttpError(400, f"Invalid command: {command}")

        else:
            raise HttpError(405, f"Method not allowed: {req.method}")

    except HttpError as error:
        status = error.status or 500
        logging.error(f"⛔ Returning error status code {status}: {str(error)}")
        res["status"] = status
        res["results"] = {
            "status": status,
            "message": str(error)
        }
        return func.HttpResponse(
            json.dumps(res),
            status_code=status,
            mimetype="application/json"
        )